Wikipedia/Code/EditFunnel.swift

```diff
@@ -0,0 +1,219 @@
+        var event: [String : Any] = ["action": action.rawValue]
+
+        if let source = source, let stringValue = source.stringValue {
+            event["source"] = stringValue
+        }
+        if let revision = revision {
+            event["revID"] = revision
+        }
+        if let editSummaryType = editSummaryType {
+            event["editSummaryTapped"] = editSummaryType.eventLoggingKey
+        }
+        if let abuseFilterName = abuseFilterName {
+            event["abuseFilterName"] = abuseFilterName
+        }
+        if let sessionToken = sessionToken {
+            event["session_token"] = sessionToken
+        }
+        if let wikidataDescriptionEdit = wikidataDescriptionEdit {
+            event["wikidataDescriptionEdit"] = wikidataDescriptionEdit.rawValue
+        }
+        if let errorText = errorText {
+            event["errorText"] = errorText
+        }
+
+        return event
+    }
+
+    override func preprocessData(_ eventData: [AnyHashable : Any]) -> [AnyHashable : Any] {
+        guard
+            let sessionID = sessionID,
+            let appInstallID = appInstallID
+        else {
+            assertionFailure("Missing required properties (sessionID or appInstallID); event won't be logged")
+            return eventData
+        }
+        guard let event = eventData as? [String: Any] else {
+            assertionFailure("Expected dictionary with keys of type String")
+            return eventData
+        }
+
+        let requiredData: [String: Any] = ["session_token": sessionID, "anon": isAnon, "app_install_id": appInstallID, "client_dt": timestamp]
+
+        return requiredData.merging(event, uniquingKeysWith: { (first, _) in first })
+    }
+
+    // MARK: Start
+
+    @objc public func logSectionEditingStartFrom(source: EditFunnelSource, revision: Int, language: String) {
+        logStart(source: source, revision: revision, language: language)
+    }
+
+    @objc public func logTitleDescriptionEditingStartFrom(source: EditFunnelSource, revision: Int, language: String) {
+        let wikidataDescriptionEdit: WikidataDescriptionEdit
+        if source == .titleDescription {
+            wikidataDescriptionEdit = .new
+        } else {
+            wikidataDescriptionEdit = .existing
+        }
+        logStart(source: source, wikidataDescriptionEdit: wikidataDescriptionEdit, revision: revision, language: language)
+    }
+
+    private func logStart(source: EditFunnelSource, wikidataDescriptionEdit: WikidataDescriptionEdit? = nil, revision: Int, language: String) {
+        // session token should be regenerated at every 'start' event
+        sessionToken = singleUseUUID()
+        log(event(action: .start, source: source, sessionToken: sessionToken, wikidataDescriptionEdit: wikidataDescriptionEdit, revision: revision), language: language)
+    }
+
+    // MARK: Onboarding
+
+    @objc public func logOnboardingPresentationInitiatedBy(source: EditFunnelSource, revision: Int, language: String) {
```

Comment:
‘logOnboardingPresentation(initiatedBy source: EditFunnelSource, revision: Int, language: String)’ may be a bit swiftier 😂 (other methods too?) not a big deal though

Reply:
like `logOnboardingPresentation(initiatedBy source: EditFunnelSource, revision: Int, language: String)` or do you have something else in mind?